keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,421 @@
+				}
+				klog.Infof("successfully added pv deletion protection finalizer to indicate the volume is to be deleted.")
+			}
+		}
+// Remaining code below skipped for brevity...
+// Please refer kubernetes/kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go for the full code.
+```
+
+When `deleteVolumeOperation` executes, the finalizer is checked, if present it implies that the PV was deleted first,
+in this scenario `DeletionTimestamp` checks can be ignored. Once the volume is deleted by the plugin the finalizer can
+be safely removed.
+
+```go
+// deleteVolumeOperation deletes a volume. This method is running in standalone
+// goroutine and already has all necessary locks.
+func (ctrl *PersistentVolumeController) deleteVolumeOperation(volume *v1.PersistentVolume) (string, error) {
+	klog.V(4).Infof("deleteVolumeOperation [%s] started", volume.Name)
+	newVolume, err := ctrl.kubeClient.CoreV1().PersistentVolumes().Get(context.TODO(), volume.Name, metav1.GetOptions{})
+	if err != nil {
+		klog.V(3).Infof("error reading persistent volume %q: %v", volume.Name, err)
+		return "", nil
+	}
+
+	// DeletionTimestamp would have been set if the PV was deleted first. Check if the PV deletion protection
+	// finalizer is set, if so, the DeletionTimestamp can be ignored
+	if !ctrl.hasPvDeletionProtectionFinalizer(volume, pvutil.PvDeletionProtectionFinalizer) {
+		if newVolume.GetDeletionTimestamp() != nil { ]
+			klog.V(3).Infof("Volume %q is already being deleted", volume.Name)
+			return "", nil
+		}
+	}
+    //.....
+    //.....
+    //.....
+    klog.V(4).Infof("deleteVolumeOperation [%s]: success", volume.Name)
+    // Remove the protection finalizer on a successful delete.
+    _, err = ctrl.removePvDeletionProtectionFinalizer(volume) //<=============================== Remove the finalizer once the volume has been deleted from plugin
+    if err!= nil {
+        klog.V(3).Infof("failed to remove pv deletion protection finalizer on the " +
+            "volume %q, err: %v", volume.Name, err)
+        return pluginName, err
+    }
+// Remaining code below skipped for brevity...
+// Please refer pkg/controller/volume/persistentvolume/pv_controller.go in kubernetes/kubernetes for the full code.
+```
+
+The above changes ensure that the plugin receives the volume deletion request.
+
+On the driver side, additional checks are made whenever `DeletionTimestamp` is set.
+
+```go
+// shouldDelete returns whether a volume should have its backing volume
+// deleted, i.e. whether a Delete is "desired"
+func (ctrl *ProvisionController) shouldDelete(ctx context.Context, volume *v1.PersistentVolume) bool {
+	if deletionGuard, ok := ctrl.provisioner.(DeletionGuard); ok {
+		if !deletionGuard.ShouldDelete(ctx, volume) {
+			return false
+		}
+	}
+
+	// In 1.9+ PV protection means the object will exist briefly with a
+	// deletion timestamp even after our successful Delete. Ignore it.
+	if ctrl.kubeVersion.AtLeast(utilversion.MustParseSemantic("v1.9.0")) {
```

Comment:
I think we can safely remove check for Kubernetes 1.9.0, which will simplify the code a bit.